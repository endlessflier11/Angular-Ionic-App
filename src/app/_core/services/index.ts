export * from './auth/auth.service';
export * from './sso/sso.service';
