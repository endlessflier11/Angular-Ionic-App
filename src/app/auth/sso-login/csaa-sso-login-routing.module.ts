import { NgModule } from '@angular/core';
import { SsoLoginPage } from './sso-login.page';
import { RouterModule } from '@angular/router';
import { AppRoutes } from '@csaadigital/mobile-mypolicy';
// This is here because the main app depends on micro-app's HttpService which requires that AppEndpoints are loaded.
// Without it, we'd have to manually fetch service locator data.
// In production, parent app does not need micro-app's HttpService.
import { BootstrapGuard } from '../../micro-apps/csaa-mobile/_core/guards/bootstrap/bootstrap.guard';

const routes: AppRoutes = [
  {
    isGroupIndex: true,
    path: '',
    component: SsoLoginPage,
    canActivate: [BootstrapGuard],
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class CsaaSsoLoginRoutingModule {}
