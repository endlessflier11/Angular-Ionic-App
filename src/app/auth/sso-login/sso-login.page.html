<csaa-loading [visible]="isLoading$|async"></csaa-loading>
<ion-content [fullscreen]="true">
  <div>
    <h2>SSO Bypass</h2>
    <ion-grid>
      <form [formGroup]="form" (submit)="submit()">
        <ion-row>
          <ion-col>
            <ion-label position="fixed">Email</ion-label>
            <ion-input
              type="email"
              placeholder="example@email.com"
              name="email"
              formControlName="email"
              data-test="email"
              (keyup.enter)="submit()"
            ></ion-input>
          </ion-col>
        </ion-row>
        <ion-row>
          <ion-col class="ion-text-center">
            <ion-button expand="block" type="submit" data-test="login" [disabled]="form.disabled"
              >Login</ion-button
            >
          </ion-col>
        </ion-row>
      </form>
      <ion-row>
        <ion-col class="ion-text-center">
          <ion-button (click)="goBack()" fill="none" data-test="back" [disabled]="form.disabled"
            >Back</ion-button
          >
        </ion-col>
      </ion-row>
      <ion-row *ngIf="error">
        <ion-col>
          <h2 class="ion-text-center" data-test="error">Login failed</h2>
          <ion-label color="danger" data-test="error-message">{{error}}</ion-label>
        </ion-col>
      </ion-row>
      <ion-row *ngIf="(translatorError$ | async)">
        <ion-col>
          <h2 class="ion-text-center" data-test="error">Translator Service error</h2>
          <ion-label color="danger" data-test="error-message"
            >{{ (translatorError$ | async)}}</ion-label
          >
        </ion-col>
      </ion-row>
    </ion-grid>
  </div>
</ion-content>
