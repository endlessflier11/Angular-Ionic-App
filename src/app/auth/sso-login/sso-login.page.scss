ion-content {
  > div {
    position: absolute;
    left: 0;
    right: 0;
    top: 50%;
    transform: translateY(-50%);
    h2 {
      text-align: center;
    }
  }

  ion-grid {
    max-width: 80%;
    margin-left: auto;
    margin-right: auto;
  }

  ion-button {
    margin-top: 20px;
    text-transform: unset;
    font-weight: bold;
  }
  ion-toggle {
    margin-right: 10px;
  }
  .ion-color-danger {
    max-height: 142px;
    display: block;
    overflow: auto;
  }
}
