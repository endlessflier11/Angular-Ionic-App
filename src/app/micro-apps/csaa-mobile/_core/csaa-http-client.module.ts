import { NgModule } from '@angular/core';
import { AngularHttpAdapter } from './services/http/adapters/angular-http.adapter';
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { IonicHttpAdapter } from './services/http/adapters/ionic-http.adapter';
import { HTTP } from '@ionic-native/http/ngx';
import { UNI_HTTP_ADAPTER } from './services/http/uni-http.model';

@NgModule({
  imports: [HttpClientModule],
  providers: [
    HTTP,
    { provide: UNI_HTTP_ADAPTER, multi: true, useClass: AngularHttpAdapter, deps: [HttpClient] },
    {
      provide: UNI_HTTP_ADAPTER,
      multi: true,
      useClass: IonicHttpAdapter,
      deps: [HTTP],
    },
  ],
  exports: [HttpClientModule],
})
export class CsaaHttpClientModule {}
