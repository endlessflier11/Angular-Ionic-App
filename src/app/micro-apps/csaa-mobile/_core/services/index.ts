export * from './router/router.service';
export * from './config/config.service';
export * from './auth/auth.service';
export * from './sso/sso.service';
export * from './storage/storage.service';
export * from './http/http.service';
export * from './analytics.service';
export * from './policy.service';
export * from './payment.service';
export * from './global-state.service';
export * from './call.service';
export * from './metadata.service';
export * from './pdf-display.service';
export * from './webview/webview.service';
