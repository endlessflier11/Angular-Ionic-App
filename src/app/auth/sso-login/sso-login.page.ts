/* eslint-disable @typescript-eslint/naming-convention,no-underscore-dangle */
import { Component, OnDestroy, OnInit } from '@angular/core';
import { NavController } from '@ionic/angular';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { BehaviorSubject, combineLatest, noop, Observable, of, Subscription } from 'rxjs';
import { AuthService, SsoService } from '../../_core/services';
import { ClubConfigHelper } from '../../_core/helpers/club-config.helper';
import { Select, Store } from '@ngxs/store';
import { ConfigState } from '../../micro-apps/csaa-mobile/_core/store/states/config.state';
import { unsubscribeIfPresent } from '../../micro-apps/csaa-mobile/_core/helpers';
import {
  FetchState,
  FetchStateError,
} from '../../micro-apps/csaa-mobile/_core/store/states/fetch.state';
import { CsaaAuthAction } from '../../micro-apps/csaa-mobile/_core/store/actions';
import { map } from 'rxjs/operators';
import {
  filterFetchErrorsBefore,
  onceTruthy,
  trackIsActive,
} from '../../micro-apps/csaa-mobile/_core/operators';

@Component({
  selector: 'app-csaa-sso-login',
  templateUrl: './sso-login.page.html',
  styleUrls: ['./sso-login.page.scss'],
})
export class SsoLoginPage implements OnInit, OnDestroy {
  public readonly form: FormGroup;
  valueChanges: Subscription;
  disableOnLoading: Subscription;

  error = null;
  @Select(FetchState.activeErrorFor(CsaaAuthAction.RequestIgAccessToken))
  _translatorError$: Observable<FetchStateError>;
  translatorError$: Observable<string>;
  @Select(FetchState.isLoading(CsaaAuthAction.RequestIgAccessToken))
  _isLoading$: Observable<boolean>;
  formSubmitionSubject = new BehaviorSubject<boolean>(false);
  isLoading$: Observable<boolean>;

  constructor(
    private readonly formBuilder: FormBuilder,
    private readonly ssoService: SsoService,
    private readonly navController: NavController,
    private readonly authService: AuthService,
    private readonly store: Store
  ) {
    this.form = this.formBuilder.group({
      email: ['', [Validators.required, Validators.email]],
    });
  }

  ngOnInit() {
    this.translatorError$ = this._translatorError$.pipe(filterFetchErrorsBefore(Date.now()));

    // This is a way to combine the state of two async and inependent actions:
    // SSO Login and CsaaAuthAction.RequestIgAccessToken
    this.isLoading$ = combineLatest([this._isLoading$, this.formSubmitionSubject]).pipe(
      map((areActionsLoading) => areActionsLoading.find(Boolean))
    );
    // so we can properly use the loading state to control form state
    this.disableOnLoading = this.isLoading$.subscribe((loading) => {
      if (loading) {
        this.form.disable();
      } else {
        this.form.enable();
      }
    });
  }

  ngOnDestroy() {
    unsubscribeIfPresent(this.valueChanges);
    unsubscribeIfPresent(this.disableOnLoading);
  }

  submit() {
    this.error = null;
    const inputs = this.form.value;
    if (this.form.invalid) {
      return;
    }

    this.ssoService
      .loginViaSsoBypass(inputs.email)
      .pipe(trackIsActive(this.formSubmitionSubject))
      .subscribe(
        () => {
          this.authService.ready().subscribe(() => {
            if (this.authService.isAuthenticated) {
              this.goToClubHome();
            } else if (
              this.store.selectSnapshot(FetchState.isLoading(CsaaAuthAction.RequestIgAccessToken))
            ) {
              this.isLoading$
                .pipe(
                  map((loading) => !loading),
                  onceTruthy()
                )
                .subscribe(() => {
                  if (
                    this.store.selectSnapshot(
                      FetchState.succeeded(CsaaAuthAction.RequestIgAccessToken)
                    )
                  ) {
                    this.goToClubHome();
                  }
                });
            }
          });
        },
        (error) => {
          console.log('CSAA:', { error });
          this.error = error.message;
          return of(error);
        }
      );
  }

  goToClubHome() {
    const theme = this.store.selectSnapshot(ConfigState.theme);
    const { clubPath } = ClubConfigHelper.getPaths(theme);
    this.navController.navigateForward(clubPath).then(noop);
  }

  goBack() {
    this.navController.back();
  }
}
