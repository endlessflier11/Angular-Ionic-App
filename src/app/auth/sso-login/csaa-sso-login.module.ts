import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IonicModule } from '@ionic/angular';
import { SsoLoginPage } from './sso-login.page';
import { ReactiveFormsModule } from '@angular/forms';
import { CsaaSsoLoginRoutingModule } from './csaa-sso-login-routing.module';
import { UiKitsModule } from '../../micro-apps/csaa-mobile/_core/ui-kits/ui-kits.module';

@NgModule({
  declarations: [SsoLoginPage],
  imports: [
    CommonModule,
    IonicModule,
    CsaaSsoLoginRoutingModule,
    ReactiveFormsModule,
    UiKitsModule,
  ],
})
export class CsaaSsoLoginModule {}
