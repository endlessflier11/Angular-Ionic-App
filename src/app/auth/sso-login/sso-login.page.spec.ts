import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { SsoLoginPage } from './sso-login.page';
import { AuthMockService, PageTestingModule, SsoMockService } from '@app/testing';
import { AuthService, SsoService } from '../../_core/services';
import { UiKitsModule } from '../../micro-apps/csaa-mobile/_core/ui-kits/ui-kits.module';

describe('SsoLoginPage', () => {
  let component: SsoLoginPage;
  let fixture: ComponentFixture<SsoLoginPage>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [SsoLoginPage],
      imports: [PageTestingModule.withConfig({ providesStorage: true }), UiKitsModule],
      providers: [
        { provide: AuthService, useClass: AuthMockService },
        { provide: SsoService, useClass: SsoMockService },
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(SsoLoginPage);
    component = fixture.componentInstance;
    fixture.detectChanges();
  }));

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it.skip('should match snapshot', async () => {
    await fixture.whenStable();
    expect(fixture).toMatchSnapshot();
  });
});
