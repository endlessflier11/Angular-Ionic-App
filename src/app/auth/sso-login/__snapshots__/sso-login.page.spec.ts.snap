// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SsoLoginPage should match snapshot 1`] = `
<app-csaa-sso-login
  authService={[Function AuthMockService]}
  form={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  navController={[Function NavController]}
  route={[Function Function]}
  ssoService={[Function SsoMockService]}
>
  <ion-content>
    <div>
      <ion-grid>
        <ion-row>
          <ion-col>
            <form
              class="ng-untouched ng-pristine ng-invalid"
              novalidate=""
            >
              <ion-label
                position="fixed"
              >
                Email
              </ion-label>
              <ion-input
                class="ng-untouched ng-pristine ng-invalid ion-untouched ion-pristine ion-invalid"
                formcontrolname="email"
                name="email"
                placeholder="example@email.com"
                type="email"
              />
              <ion-label
                position="fixed"
              >
                Password
              </ion-label>
              <ion-input
                class="ng-untouched ng-pristine ng-invalid ion-untouched ion-pristine ion-invalid"
                formcontrolname="password"
                name="password"
                placeholder="password"
                type="password"
              />
              <ion-button
                expand="block"
                type="submit"
              >
                Log In
              </ion-button>
            </form>
          </ion-col>
        </ion-row>
        <ion-row>
          <ion-col
            class="ion-text-center"
          >
            <ion-button
              fill="none"
            >
              Back
            </ion-button>
          </ion-col>
        </ion-row>
      </ion-grid>
    </div>
  </ion-content>
</app-csaa-sso-login>
`;
